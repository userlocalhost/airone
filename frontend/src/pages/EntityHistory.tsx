import {
  Box,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
  Typography,
} from "@mui/material";
import React, { FC } from "react";
import { Link, useParams } from "react-router-dom";
import { useAsync } from "react-use";

import { entitiesPath, topPath } from "../Routes";
import { AironeBreadcrumbs } from "../components/common/AironeBreadcrumbs";
import { Loading } from "../components/common/Loading";
import { getEntityHistory } from "../utils/AironeAPIClient";

const Operations = {
  ADD: 1 << 0,
  MOD: 1 << 1,
  DEL: 1 << 2,
};

const Targets = {
  ENTITY: 1 << 3,
  ATTR: 1 << 4,
  ENTRY: 1 << 5,
};

const TargetOperation = {
  ADD_ENTITY: Operations.ADD + Targets.ENTITY,
  ADD_ATTR: Operations.ADD + Targets.ATTR,
  MOD_ENTITY: Operations.MOD + Targets.ENTITY,
  MOD_ATTR: Operations.MOD + Targets.ATTR,
  DEL_ENTITY: Operations.DEL + Targets.ENTITY,
  DEL_ATTR: Operations.DEL + Targets.ATTR,
  DEL_ENTRY: Operations.DEL + Targets.ENTRY,
};

export const EntityHistory: FC = () => {
  const { entityId } = useParams<{ entityId: number }>();

  const history = useAsync(async () => {
    const resp = await getEntityHistory(entityId);
    return await resp.json();
  });

  return (
    <Box className="container">
      <AironeBreadcrumbs>
        <Typography component={Link} to={topPath()}>
          Top
        </Typography>
        <Typography component={Link} to={entitiesPath()}>
          エンティティ一覧
        </Typography>
        <Typography color="textPrimary">変更履歴</Typography>
      </AironeBreadcrumbs>

      <Table className="table">
        <TableHead>
          <TableRow>
            <TableCell>Operator</TableCell>
            <TableCell>Operation</TableCell>
            <TableCell>Details</TableCell>
            <TableCell>Time</TableCell>
          </TableRow>
        </TableHead>

        <TableBody>
          {history.loading ? (
            <Loading />
          ) : (
            <>
              {history.value.map((column, index) => (
                <TableRow key={index}>
                  <TableCell>{column.user.username}</TableCell>
                  <TableCell>
                    {(() => {
                      switch (column.operation) {
                        case TargetOperation.ADD_ENTITY:
                          return <Typography>作成</Typography>;
                        case TargetOperation.MOD_ENTITY:
                          return <Typography>変更</Typography>;
                        case TargetOperation.DEL_ENTITY:
                          return <Typography>削除</Typography>;
                        default:
                          return (
                            <Typography>
                              {column.operation} ({TargetOperation.ADD_ENTITY})
                            </Typography>
                          );
                      }
                    })()}
                  </TableCell>
                  <TableCell>
                    <Table>
                      <TableBody>
                        {column.details.map((detail, index) => (
                          <TableRow key={index}>
                            <TableCell>
                              {(() => {
                                switch (detail.operation) {
                                  case TargetOperation.MOD_ENTITY:
                                    return <Typography>変更</Typography>;
                                  case TargetOperation.ADD_ATTR:
                                    return <Typography>属性追加</Typography>;
                                  case TargetOperation.MOD_ATTR:
                                    return <Typography>属性変更</Typography>;
                                  case TargetOperation.DEL_ATTR:
                                    return <Typography>属性削除</Typography>;
                                  default:
                                    return <Typography />;
                                }
                              })()}
                            </TableCell>
                            <TableCell>{detail.target_obj.name}</TableCell>
                            <TableCell>{detail.text}</TableCell>
                          </TableRow>
                        ))}
                      </TableBody>
                    </Table>
                  </TableCell>
                  <TableCell>{column.time}</TableCell>
                </TableRow>
              ))}
            </>
          )}
        </TableBody>
      </Table>
    </Box>
  );
};
