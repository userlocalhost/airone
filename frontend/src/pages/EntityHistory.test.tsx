/**
 * @jest-environment jsdom
 */

import {
  render,
  waitForElementToBeRemoved,
  screen,
} from "@testing-library/react";
import React from "react";

import { TestWrapper } from "../utils/TestWrapper";

import { EntityHistory } from "./EntityHistory";

afterEach(() => {
  jest.clearAllMocks();
});

test("should match snapshot", async () => {
  const entityHistory = [
    {
      operation: 1,
      time: "2022-01-01 00:00:00",
      user: {
        username: "aaa",
      },
      details: [
        {
          operation: 1,
          text: "aaa",
          target_obj: {
            name: "aaa",
          },
        },
      ],
    },
    {
      operation: 2,
      time: "2022-01-01 00:00:00",
      user: {
        username: "bbb",
      },
      details: [
        {
          operation: 2,
          text: "bbb",
          target_obj: {
            name: "bbb",
          },
        },
      ],
    },
  ];

  /* eslint-disable */
  jest
    .spyOn(require("../utils/AironeAPIClient"), "getEntityHistory")
    .mockResolvedValue({
      json() {
        return Promise.resolve(entityHistory);
      },
    });
  /* eslint-enable */

  // wait async calls and get rendered fragment
  const result = render(<EntityHistory />, {
    wrapper: TestWrapper,
  });
  await waitForElementToBeRemoved(screen.getByTestId("loading"));

  expect(result).toMatchSnapshot();
});
