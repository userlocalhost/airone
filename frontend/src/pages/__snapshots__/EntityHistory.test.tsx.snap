// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container MuiBox-root css-0"
      >
        <nav
          aria-label="breadcrumb"
          class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root makeStyles-breadcrumbs-1 css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
        >
          <ol
            class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
          >
            <li
              class="MuiBreadcrumbs-li"
            >
              <a
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                href="/new-ui/"
              >
                Top
              </a>
            </li>
            <li
              aria-hidden="true"
              class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
            >
              /
            </li>
            <li
              class="MuiBreadcrumbs-li"
            >
              <a
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                href="/new-ui/entities"
              >
                エンティティ一覧
              </a>
            </li>
            <li
              aria-hidden="true"
              class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
            >
              /
            </li>
            <li
              class="MuiBreadcrumbs-li"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1gqyz35-MuiTypography-root"
              >
                変更履歴
              </p>
            </li>
          </ol>
        </nav>
        <table
          class="MuiTable-root table css-rqglhn-MuiTable-root"
        >
          <thead
            class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              >
                Operator
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              >
                Operation
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              >
                Details
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              >
                Time
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                aaa
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  1
                   (
                  9
                  )
                </p>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                <table
                  class="MuiTable-root css-rqglhn-MuiTable-root"
                >
                  <tbody
                    class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
                  >
                    <tr
                      class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
                    >
                      <td
                        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                        />
                      </td>
                      <td
                        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                      >
                        aaa
                      </td>
                      <td
                        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                      >
                        aaa
                      </td>
                    </tr>
                  </tbody>
                </table>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                2022-01-01 00:00:00
              </td>
            </tr>
            <tr
              class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                bbb
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  2
                   (
                  9
                  )
                </p>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                <table
                  class="MuiTable-root css-rqglhn-MuiTable-root"
                >
                  <tbody
                    class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
                  >
                    <tr
                      class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
                    >
                      <td
                        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                        />
                      </td>
                      <td
                        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                      >
                        bbb
                      </td>
                      <td
                        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                      >
                        bbb
                      </td>
                    </tr>
                  </tbody>
                </table>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                2022-01-01 00:00:00
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container MuiBox-root css-0"
    >
      <nav
        aria-label="breadcrumb"
        class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root makeStyles-breadcrumbs-1 css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
      >
        <ol
          class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
        >
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              href="/new-ui/"
            >
              Top
            </a>
          </li>
          <li
            aria-hidden="true"
            class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
          >
            /
          </li>
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              href="/new-ui/entities"
            >
              エンティティ一覧
            </a>
          </li>
          <li
            aria-hidden="true"
            class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
          >
            /
          </li>
          <li
            class="MuiBreadcrumbs-li"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1gqyz35-MuiTypography-root"
            >
              変更履歴
            </p>
          </li>
        </ol>
      </nav>
      <table
        class="MuiTable-root table css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Operator
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Operation
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Details
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Time
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              aaa
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                1
                 (
                9
                )
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              <table
                class="MuiTable-root css-rqglhn-MuiTable-root"
              >
                <tbody
                  class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
                >
                  <tr
                    class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                      />
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                    >
                      aaa
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                    >
                      aaa
                    </td>
                  </tr>
                </tbody>
              </table>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              2022-01-01 00:00:00
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              bbb
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                2
                 (
                9
                )
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              <table
                class="MuiTable-root css-rqglhn-MuiTable-root"
              >
                <tbody
                  class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
                >
                  <tr
                    class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                      />
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                    >
                      bbb
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                    >
                      bbb
                    </td>
                  </tr>
                </tbody>
              </table>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              2022-01-01 00:00:00
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
